src/DotNetty.Transport/Channels/Sockets/SocketDatagramChannel.cs

```diff
@@ -0,0 +1,470 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Transport.Channels.Sockets
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Contracts;
+    using System.Net;
+    using System.Net.NetworkInformation;
+    using System.Net.Sockets;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using DotNetty.Buffers;
+    using DotNetty.Common;
+    using DotNetty.Common.Concurrency;
+    using DotNetty.Common.Internal.Logging;
+    using DotNetty.Common.Utilities;
+
+    public class SocketDatagramChannel : AbstractSocketMessageChannel, IDatagramChannel
+    {
+        static readonly IInternalLogger Logger = InternalLoggerFactory.GetInstance<SocketDatagramChannel>();
+        static readonly Action<object, object> ReceiveFromCompletedSyncCallback = OnReceiveFromCompletedSync;
+        static readonly ChannelMetadata ChannelMetadata = new ChannelMetadata(true);
+
+        readonly DefaultDatagramChannelConfig config;
+        readonly IPEndPoint anyRemoteEndPoint;
+
+        public SocketDatagramChannel()
+            : this(new Socket(SocketType.Dgram, ProtocolType.Udp))
+        {
+        }
+
+        public SocketDatagramChannel(AddressFamily addressFamily)
+            : this(new Socket(addressFamily, SocketType.Dgram, ProtocolType.Udp))
+        {
+        }
+
+        public SocketDatagramChannel(Socket socket)
+            : base(null, socket)
+        {
+            this.config = new DefaultDatagramChannelConfig(this, socket);
+            this.anyRemoteEndPoint = new IPEndPoint(
+                socket.AddressFamily == AddressFamily.InterNetwork ? IPAddress.Any : IPAddress.IPv6Any,
+                IPEndPoint.MinPort);
+        }
+
+        public override IChannelConfiguration Configuration => this.config;
+
+        public override ChannelMetadata Metadata => ChannelMetadata;
+
+        protected override EndPoint LocalAddressInternal => this.Socket.LocalEndPoint;
+
+        protected override EndPoint RemoteAddressInternal => this.Socket.RemoteEndPoint;
+
+        protected override void DoBind(EndPoint localAddress)
+        {
+            this.Socket.Bind(localAddress);
+            this.CacheLocalAddress();
+
+            this.SetState(StateFlags.Active);
+        }
+
+        public override bool Active => 
```

Comment:
from commit in netty adding "active_on_registration" option:
> Introduce a new ChannelOption called DATAGRAM_CHANNEL_ACTIVE_ON_REGISTRATION. Related to [#1830]  This ChannelOption allows to tell the DatagramChannel implementation to be active as soon as they are registrated to their EventLoop. This can be used to make it possible to write to a not bound DatagramChannel. The ChannelOption is marked as @deprecated as I'm looking for a better solution in master which breaks default behaviour with 4.0 branch.```